### PR TITLE
feat: separate discount accounting for buying and selling

### PR DESCRIPTION
**Pull request by rahib-hassan**
Discount Accounting removed from Account Settings for separation in Sales and Purchase Invoice and added to Selling and Buying Settings.

Removed hide/un-hide discount account in Item since default_discount_account has been moved to Item Defaults.

Docs updated: https://docs.erpnext.com/docs/v13/user/manual/en/accounts/articles/discount_accounting

### CODE REVIEW

**Comment by ankush** [comment]
@rahib-hassan follow semantic commit convention

https://www.conventionalcommits.org/en/v1.0.0/ 


You can amend and repush you last commit.

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.py [hunk 1]
```diff
@@ -52,74 +51,6 @@ def enable_payment_schedule_in_print(self):
 				validate_fields_for_doctype=False,
 			)
 
-	def toggle_discount_accounting_fields(self):
-		enable_discount_accounting = cint(self.enable_discount_accounting)
```
Comment (ankush) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.py:56]: These fields should be removed from accounts settings.

**Review by ankush — CHANGES_REQUESTED** [review]
A patch should be added for users who have already enabled it globally to enable it in both sales/purchase side?

File: erpnext/patches/v14_0/discount_accounting_separation.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import frappe
+
+def execute():
+    doc = frappe.get_doc("Accounts Settings")
+    discount_account = doc.enable_discount_accounting
```
Comment (ankush) [inline erpnext/patches/v14_0/discount_accounting_separation.py]: for patches this is fine but avoid getting whole doc just for single value. `get_single_value` or `get_value` should be used. 

reason: `get_doc` is bad for performance.

**Comment by ankush** [comment]
All dependent code needs to be updated. This is the reason behind failing tests. 


Screenshot 2022-04-12 at 6 41 25 PM

File: erpnext/patches/v14_0/discount_accounting_separation.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    if discount_account:
+        for doctype in ["Buying Settings", "Selling Settings"]:
```
Comment (ankush) [inline erpnext/patches/v14_0/discount_accounting_separation.py]:
```suggestion
        for doctype in ["Buying Settings", "Selling Settings"]:
```

We use tabs for formatting code. 

Setup pre-commit and add editorconfig plugin to automate this: https://github.com/frappe/erpnext/wiki/Pull-Request-Checklist#linting

File: erpnext/buying/doctype/buying_settings/buying_settings.json [hunk 4]
```diff
@@ -133,14 +134,21 @@
   {
    "fieldname": "column_break_12",
    "fieldtype": "Column Break"
+  },
+  {
+   "default": "0",
+   "description": "If enabled, additional ledger entries will be made for discounts in a separate Discount Account",
+   "fieldname": "enable_discount_accounting",
+   "fieldtype": "Check",
+   "label": "Enable Discount Accounting"
```
Comment (deepeshgarg007) [inline erpnext/buying/doctype/buying_settings/buying_settings.json]:
```suggestion
   "label": "Enable Discount Accounting for Buying"
```

File: erpnext/selling/doctype/selling_settings/selling_settings.json [hunk 5]
```diff
@@ -164,14 +165,21 @@
    "fieldname": "editable_bundle_item_rates",
    "fieldtype": "Check",
    "label": "Calculate Product Bundle Price based on Child Items' Rates"
+  },
+  {
+   "default": "0",
+   "description": "If enabled, additional ledger entries will be made for discounts in a separate Discount Account",
+   "fieldname": "enable_discount_accounting",
+   "fieldtype": "Check",
+   "label": "Enable Discount Accounting"
```
Comment (deepeshgarg007) [inline erpnext/selling/doctype/selling_settings/selling_settings.json]:
```suggestion
   "label": "Enable Discount Accounting for Sales"
```

**Review by deepeshgarg007 — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
@rahib-hassan update docs and link it in PR description then the docs check will pass: https://docs.erpnext.com/docs/v13/user/manual/en/accounts/articles/discount_accounting

File: erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py [hunk 6]
```diff
@@ -3181,6 +3175,12 @@ def test_gain_loss_with_advance_entry(self):
 		)
 
 
+def enable_discount_accounting(enable=1):
```
Comment (ankush) [inline erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py]: There's already a simpler/reliable way to do this. 

Example: 
https://github.com/frappe/erpnext/blob/c42dcaa6cff5c87a10d58854cd10089f5ba53efb/erpnext/stock/doctype/stock_entry/test_stock_entry.py#L769

File: erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py [hunk 7]
```diff
@@ -2702,6 +2702,7 @@ def test_sales_invoice_with_discount_accounting_enabled(self):
 		]
 
 		check_gl_entries(self, si.name, expected_gle, add_days(nowdate(), -1))
+		enable_discount_accounting(enable=0)
```
Comment (ankush) [inline erpnext/accounts/doctype/sales_invoice/test_sales_invoice.py]: @rahib-hassan this isn't required, `change_settings` decorator rolls it back to previous value after test runs.

**Comment by rahib-hassan** [comment]
> A patch should be added for users who have already enabled it globally to enable it in both sales/purchase side?

@ankush This has been done already.

**Comment by ankush** [comment]
@deepeshgarg007 review? I am not much familiar with this but based on diff it appears to be okay to me.

**Comment by deepeshgarg007** [comment]
@Mergifyio backport version-13-hotfix

**Comment by ankush** [comment]
@rahib-hassan backport this manually on v13 if required. 

because of merge commit it's better to backport this as "squashed"


```bash
git checkout -b discount_account_v13 version-13-hotfix
wget https://github.com/frappe/erpnext/pull/30642.patch
git apply 30642.patch
# commit ... pr.. 
```

**Pull request merged**